### ISSUE TITLE
Few-Shot and Continual Learning With Attentive Independent Mechanisms

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#142]
## TL; DR

- Train independent *experts* (mechanisms) that refine representations upon similarity-based retrieval.

## Motivation

- To learn fast and without forgetting, we need task-agnostic representation learner and adequate feature refiner.

## Related Work

*Personal Opinion*
- RAG (Retrieval Augmented Generation): For solving generative language modeling tasks, employ stored document representations using attention-based retrieval.
- YOLOR (You Only Learn One Representation): Keep separate set of implicit knowledge (*analogous to subconscious level knowledge in human cognitive system*) and leverage them for handling multiple tasks with a single model

### Attentive Independent Mechanisms

![image](https://user-images.githubusercontent.com/46308407/147636461-0fa17b62-4e48-48fd-9bef-b4b8323fa9f3.png)

![image](https://user-images.githubusercontent.com/46308407/147636891-484e1f15-dfe9-4ec2-9bd6-771e512cc1dc.png)
![image](https://user-images.githubusercontent.com/46308407/147636900-6195a5ee-f888-462d-ab99-ffb1fc4ded1e.png)
![image](https://user-images.githubusercontent.com/46308407/147636910-b74758da-6fbf-4f73-992a-25004f6381c6.png)
![image](https://user-images.githubusercontent.com/46308407/147636922-7084dfe4-fa0a-4614-a464-2902f3b42508.png)

- Insert an additional set of mechanisms in-between the feature extractor and the prediction head.
- For a given representation, retrieve top-K mechanisms (*experts, reference knowledge*) via attention and refine the *task-agnostic* feature for specific task at hand. 
- The idea (*in my opinion*) is to disentangle task-agnostic universal representation and task-specific knowledge.
- Refine the universal features with ensemble of mechanisms to suit for the downstream task.

### Stochastic Sampling

- The proposed method only uses top-k most relevant mechanisms for the task, and therefore only these k mechanisms are updated using downstream gradient.
- Potential risk that only a fixed set of mechanisms (k, in the worst case) are used throughout the whole process.
- To encourage the usage of diverse mechanisms for different instances and tasks, apply stochastic sampling.
- First sample top (k+l) samples upon similarity, then choose k out of (k+l) uniformly.

### Training / Testing Procedure

![image](https://user-images.githubusercontent.com/46308407/147636956-20461b95-4494-4c50-b233-9d7a51ca394c.png)

### Application: Few-shot

![image](https://user-images.githubusercontent.com/46308407/147636974-964da69b-5094-4d60-a462-b95fdef49893.png)

- Feature extractor is frozen during training.
- Fast Weight = Inner Loop, Slow Weight = Outer Loop

### Application: Continual Learning

![image](https://user-images.githubusercontent.com/46308407/147637187-9686dcf9-e0d1-416c-b8de-06429a74cb88.png)

- For OML, feature extractor is updated in the outer loop, and independent mechanisms are updated with the classification head as fast weights (inner loop).

### Implementation Details

- 32 mechanisms with top-8 selection, slack=2 (retrieve 8+2 and uniformly choose 8)

![image](https://user-images.githubusercontent.com/46308407/147637366-91e202cd-8496-4f23-a788-09ca24e98510.png)

- We can see that not all mechanisms are well employed. Some mechanisms are used across all classes (white vertical lines) and some are almost never used (dark blue / black vertical lines).
- However, there are *some* mechanisms that are turned on and off depending on the class of the input. This suggests that although there are some technical improvements to be made (especially regarding sampling or hard top-k selection), this framework does provide certain level of flexibility for modeling diverse tasks.

![image](https://user-images.githubusercontent.com/46308407/147637640-40bbe4b2-306a-44f1-87f4-fdd24c8f505a.png)
![image](https://user-images.githubusercontent.com/46308407/147637695-d92658e4-4896-4c82-86a4-1ee6e7800a54.png)
![image](https://user-images.githubusercontent.com/46308407/147637720-c9bbf6b4-e256-4ef5-ae9c-bdccb14427e6.png)

## Comments

- This work in itself does display some blind spots for further improvement, I believe the future direction it points at is very promising and well-declared.
- Although this work doesn't use specific phrases like *external knowledge* or *implicit knowledge*, I think the high-level idea is quite similar to that of RAG or YOLOR.
- AIM (Attentive Independent Mechanisms) also keeps certain number of mechanisms (composed of hidden states for retrieval and weights for feature refinement) and retrieve most relevant components for each input instance.
- Especially for problem settings like continual learning, I think encoding the whole knowledge into model parameters is not an efficient solution. We know, for one, that over-parameterized DNNs form a connected manifold of local minima, within which the model parameters can move freely even upon convergence.
- Hence, continual learning, in its very own definition, is quite an unstable problem, as we are almost never sure from which point we are starting from.
- Plus, directly combatting catastrophic forgetting in terms of backward transfer (in my very personal opinion) can be a hopeless endeavor, desperately striving to avoid the worst instead of directly shooting for the best.
- Hence, I welcome such works that try to enlarge the span of our conventional framework and incorporate diverse *new* concepts, such as common knowledge, retrieval or *independent set of attentive experts*.
